test/cases/testdata/withkeyword/test-with-builtin-mock.yaml

```diff
@@ -0,0 +1,391 @@
+  query: data.test.p = x
+  want_result:
+  - x: true
+- data:
+  modules:
+  - |
+    package test
+
+    f(_) = 1
+    g(x) = count(x) # replaced with f by inner "with"
+
+    q = y {
+      y = count([1,2,3]) with count as f
+    }
+    p = y {
+      y = q with count as g
+    }
+  note: 'withkeyword/builtin: multiple with'
+  query: data.test.p = x
+  want_result:
+  - x: 1
+- data:
+  modules:
+  - |
+    package test
+
+    f(x) = object.union_n(x) { { x: i | x := ["a", "a"][i]} } # never called, runtime error
+    g(x) = count(x)
+    p {
+      q with count as f
+    }
+    q {
+      r with object.union_n as g
+    }
+    r {
+      object.union_n([{}])
+    }
+  note: 'withkeyword/builtin: mock will not call other mock'
+  query: data.test.p = x
+  want_result:
+  - x: true
+- data:
+  modules:
+  - |
+    package test
+
+    f(x) = {"foo": 3} {
+      object.union_n(x) # b1
+      true
+    }
+    g(x) = 123 {
+      count(x) # b2
+      s with array.reverse as h
+    }
+    h(_) = ["replaced"]
+    p { q with object.union_n as f }
+    q { r with count as g }
+    r { x := [{}]; count(x) == 123; object.union_n(x) == {"foo": 3} }
+    s { x := [{}]; array.reverse(x) == ["replaced"] }
+  note: 'withkeyword/builtin: nested scope handling'
+  query: data.test.p = x
+  want_result:
+  - x: true
+- data:
+  modules:
+  - |
+    package test
+
+    p = y {
+      y = time.now_ns() with time.now_ns as 12300
+    }
+  note: 'withkeyword/builtin-value: arity-0, captured output'
+  query: data.test.p = x
+  want_result:
+  - x: 12300
+- data:
+  modules:
+  - |
+    package test
+
+    p {
+      time.now_ns() with time.now_ns as false
+    }
+  note: 'withkeyword/builtin-value: arity-0, false'
+  query: data.test.p = x
+  want_result: []
+- data:
+  modules:
+  - |
+    package test
+
+    p {
+      time.now_ns() with time.now_ns as true
+    }
+  note: 'withkeyword/builtin-value: arity-0'
+  query: data.test.p = x
+  want_result:
+  - x: true
+- data:
+  modules:
+  - |
+    package test
+
+    p {
+      x := true
+      time.now_ns() with time.now_ns as x
+    }
+  note: 'withkeyword/builtin-value: arity-0, var'
+  query: data.test.p = x
+  want_result:
+  - x: true
+- data:
+  modules:
+  - |
+    package test
+
+    p {
+      count([]) == 1 with count as 1
+    }
+  note: 'withkeyword/builtin-value: arity-1'
+  query: data.test.p = x
+  want_result:
+  - x: true
+- data:
+  modules:
+  - |
+    package test
+
+    p {
+      count([], 1) with count as 1
+    }
+  note: 'withkeyword/builtin-value: arity-1, captured'
+  query: data.test.p = x
+  want_result:
+  - x: true
+- data:
+  modules:
+  - |
+    package test
+
+    p {
+      count(input) == 1 with count as 1
+    }
+  note: 'withkeyword/builtin-value: arity-1, input must still be defined'
+  query: data.test.p = x
+  want_result: []
+- data:
+  modules:
+  - |
+    package test
+
+    p = x {
+      x = time.now_ns() with time.now_ns as opa.runtime
```

Comment:
Yeah... but we don't have that many arity-0 built-ins. It's only time.now_ns() and opa.runtime(); since mocking rego.metadata.*() is forbidden. We could add a test-only-builtin, but that could be a burden for any other users of the test data corpus.

But I'll fix those arity-n (n > 0) cases you've commented on below :+1: